File: R/filter-samples.R
# Internal: sample-level outlier detection from per-sample status rates.
# See DESIGN.md §4.2 and decisions.md #15-17.

# Apply one of three thresholding methods to a per-sample rate vector.
# Returns a logical vector marking outliers.
.apply_outlier_method <- function(rates, method, percentile_cutoff) {
  iqr_outlier <- function(x) {
    q <- stats::quantile(x, c(0.25, 0.75), na.rm = TRUE)
    threshold <- unname(q[2L]) + 1.5 * unname(q[2L] - q[1L])
    !is.na(x) & x > threshold
  }
  pct_outlier <- function(x) {
    threshold <- unname(stats::quantile(x, percentile_cutoff, na.rm = TRUE))
    !is.na(x) & x > threshold
  }
  switch(method,
    "IQR"        = iqr_outlier(rates),
    "percentile" = pct_outlier(rates),
    "both"       = iqr_outlier(rates) | pct_outlier(rates),
    stop("Unknown outlier method: ", method, call. = FALSE)
  )
}

# Compute per-sample <LOD / >ULOQ rates and decide which samples are
# outliers. Returns a list with the logical outlier vector and a
# per-sample log frame ready to attach to PreprocessedData.
.detect_sample_outliers <- function(sample_status,
                                    metric = c("below_lod", "above_uloq", "either"),
                                    method = c("IQR", "percentile", "both"),
                                    percentile_cutoff = 0.95) {
  metric <- match.arg(metric)
  method <- match.arg(method)
  n_metab <- ncol(sample_status)
  if (n_metab == 0L) {
    stop("sample_status has zero metabolite columns.", call. = FALSE)
  }
  lod_rate  <- rowSums(sample_status == "< LOD",  na.rm = TRUE) / n_metab
  uloq_rate <- rowSums(sample_status == "> ULOQ", na.rm = TRUE) / n_metab

  outlier_lod  <- .apply_outlier_method(lod_rate, method, percentile_cutoff)
  outlier_uloq <- .apply_outlier_method(uloq_rate, method, percentile_cutoff)

  is_outlier <- switch(metric,
    "below_lod"  = outlier_lod,
    "above_uloq" = outlier_uloq,
    "either"     = outlier_lod | outlier_uloq
  )

  list(
    outliers = is_outlier,
    log = data.frame(
      lod_rate = lod_rate,
      uloq_rate = uloq_rate,
      outlier_below_lod = outlier_lod,
      outlier_above_uloq = outlier_uloq,
      is_outlier = is_outlier,
      stringsAsFactors = FALSE
    )
  )
}

File: python/metabosetr/filter_samples.py
"""Sample-level outlier detection (port of ``filter-samples.R``).

DESIGN.md §4.2. Per-sample ``<LOD`` / ``>ULOQ`` rates drive outlier
flagging via IQR / percentile / both. ``numpy`` linear quantile matches R's
default ``quantile`` type 7.
"""

from __future__ import annotations

import numpy as np
import pandas as pd


def _apply_outlier_method(rates: np.ndarray, method: str, percentile_cutoff: float) -> np.ndarray:
    rates = np.asarray(rates, dtype=float)
    finite = ~np.isnan(rates)

    def iqr_outlier(x):
        q1, q3 = np.nanquantile(x, [0.25, 0.75])  # type 7 (linear)
        threshold = q3 + 1.5 * (q3 - q1)
        return finite & (x > threshold)

    def pct_outlier(x):
        threshold = np.nanquantile(x, percentile_cutoff)
        return finite & (x > threshold)

    if method == "IQR":
        return iqr_outlier(rates)
    if method == "percentile":
        return pct_outlier(rates)
    if method == "both":
        return iqr_outlier(rates) | pct_outlier(rates)
    raise ValueError(f"Unknown outlier method: {method}")


def _detect_sample_outliers(
    sample_status: np.ndarray,
    metric: str = "below_lod",
    method: str = "IQR",
    percentile_cutoff: float = 0.95,
) -> dict:
    """Per-sample outlier flags plus a per-sample log frame."""
    if metric not in ("below_lod", "above_uloq", "either"):
        raise ValueError(f"Unknown metric: {metric}")
    status = np.asarray(sample_status, dtype=object)
    n_metab = status.shape[1]
    if n_metab == 0:
        raise ValueError("sample_status has zero metabolite columns.")

    lod_rate = (status == "< LOD").sum(axis=1) / n_metab
    uloq_rate = (status == "> ULOQ").sum(axis=1) / n_metab

    outlier_lod = _apply_outlier_method(lod_rate, method, percentile_cutoff)
    outlier_uloq = _apply_outlier_method(uloq_rate, method, percentile_cutoff)

    if metric == "below_lod":
        is_outlier = outlier_lod
    elif metric == "above_uloq":
        is_outlier = outlier_uloq
    else:
        is_outlier = outlier_lod | outlier_uloq

    log = pd.DataFrame(
        {
            "lod_rate": lod_rate,
            "uloq_rate": uloq_rate,
            "outlier_below_lod": outlier_lod,
            "outlier_above_uloq": outlier_uloq,
            "is_outlier": is_outlier,
        }
    )
    return {"outliers": np.asarray(is_outlier, dtype=bool), "log": log}
